lib/App.js

```diff
@@ -176,40 +177,111 @@ App.prototype.loadViews = function() {};
+  );
+};
+
+App.prototype.component = function(name, constructor, isDependency) {
+  if (typeof name === 'function') {
+    constructor = name;
+    name = null;
+  }
+  if (typeof constructor !== 'function') {
+    throw new Error('Missing component constructor argument');
   }
-  // DEPRECATED: constructor.prototype.view and constructor.prototype.name
-  var viewFilename = constructor.view || constructor.prototype.view;
-  var constructorIs = constructor.is || constructor.prototype.name;
 
-  // Inherit from Component
-  components.extendComponent(constructor);
+  var viewProp = constructor.view;
+  var viewIs, viewFilename, viewSource, viewDependencies;
+  // Always using an object for the static `view` property is preferred
+  if (viewProp && typeof viewProp === 'object') {
+    viewIs = viewProp.is;
+    viewFilename = viewProp.file;
+    viewSource = viewProp.source;
+    viewDependencies = viewProp.dependencies;
+  } else {
+    // Ignore other properties when `view` is an object. It is possible that
+    // properties could be inherited from a parent component when extending it.
+    //
+    // DEPRECATED: constructor.prototype.name and constructor.prototype.view
+    // use the equivalent static properties instead
+    viewIs = constructor.is || constructor.prototype.name;
+    viewFilename = constructor.view || constructor.prototype.view;
+  }
+  var viewName = name || viewIs ||
+    (viewFilename && path.basename(viewFilename, '.html'));
 
-  if (viewFilename) {
-    viewName = viewName || constructorIs || path.basename(viewFilename, '.html');
-    this.loadViews(viewFilename, viewName);
+  if (!viewName) {
+    throw new Error('No view specified for component');
+  }
+  if (viewFilename && viewSource) {
+    throw new Error('Component may not specify both a view file and source');
+  }
 
-  } else if (!viewName) {
-    if (constructorIs) {
-      viewName = constructorIs;
-      var view = this.views.register(viewName);
-      view.template = templates.emptyTemplate;
-    } else {
-      throw new Error('No view name specified for component');
+  // TODO: DRY. This is copy-pasted from derby-templates
+  var mapName = viewName.replace(/:index$/, '');
+  var currentView = this.views.nameMap[mapName];
+  var currentConstructor = (currentView && currentView.componentFactory) ?
+    currentView.componentFactory.constructor :
+    this._pendingComponentMap[mapName];
+
+  // Avoid registering the same component twice; we want to avoid the overhead
+  // of loading view files from disk again. This is also what prevents
+  // circular dependencies from infinite looping
+  if (currentConstructor === constructor) return;
+
+  // Calling app.component() overrides existing views or components. Prevent
+  // dependencies from doing this without warning
+  if (isDependency && currentView) {
+    throw new Error('Dependencies cannot override existing views. Already registered "' + viewName + '"');
```

Comment:
This line is almost always executed, thus dependencies don't really work for me. I'll add a new issue.

Reply:
See #571 :)